test/test.ml

```diff
@@ -0,0 +1,60 @@
+
+module Normal : sig 
+  type t = Foo of int | Bar | Exception [@@deriving variants]
+end = struct
+  type t = Foo of int | Bar | Exception [@@deriving variants]
+end
+
+module Normal_inline_record : sig 
+  type t =  Foo of { a : int; b : string} | Bar [@@deriving variants]
+end = struct
+  type t =  Foo of { a : int; b : string} | Bar [@@deriving variants]
+end
+
+module Poly : sig
+  type t =  [ `Foo of int | `Bar | `Exception ] [@@deriving variants]
+end= struct
+  type t =  [ `Foo of int | `Bar | `Exception ] [@@deriving variants]
+end
+
+module Wildcard : sig
+  type _ t = A | B [@@deriving variants]
+end = struct
+  type _ t = A | B [@@deriving variants]
+end
+
+module Gadt : sig 
+  type _ t = 
+    | Bool : bool -> bool t
+    | Int : int -> int t
+    | Cond : bool t * 'a t * 'a t -> 'a t
+  [@@deriving variants] 
+end =  struct
+  type _ t = 
+    | Bool : bool -> bool t
+    | Int : int -> int t
+    | Cond : bool t * 'a t * 'a t -> 'a t
+  [@@deriving variants] 
+end
+
+module Gadt_inline_record : sig 
+  type _ t = 
+    | Bool : bool -> bool t
+    | Int : int -> int t
+    | Cond : { cond: bool t; true_case: 'a t; false_case: 'a t } -> 'a t
+  [@@deriving variants] 
+end =  struct
+  type _ t = 
+    | Bool : bool -> bool t
+    | Int : int -> int t
+    | Cond : { cond: bool t; true_case: 'a t; false_case: 'a t } -> 'a t
+  [@@deriving variants] 
+end
+
+module Gadt_arity: sig 
+  type (_,_) t = Eq: ('a,'a) t
+  [@@deriving variants] 
+end =  struct
+  type (_,_) t = Eq: ('a,'a) t
+  [@@deriving variants] 
+end
```

Comment:
Can we add a test with tricky variable names: `type 'a t = A : 'a * 'b -› 'b t`, for instance.

Reply:
I tried to cover all the bases with a single GADT declaration:

```
  type _ t = 
    | Bool : bool -> bool t
    | Int : int -> int t
    | Cond : { cond: bool t ; true_branch: 'a t ; false_branch: 'a t } -> 'a t
    | Pair : 'a t * 'b t -> ('a * 'b) t
    | Swap : ('a * 'b) t -> ('b * 'a) t
    | Fst: ('a * 'b) t -> 'a t 
    | Snd : ('a * 'b) t -> 'b t
  [@@deriving variants] 
```

This brought to light an issue in the case where a type variable is existentially qualified (as in the case you proposed). For such constructors, we can't generate a `getter` and I _think_ we cannot produce `map` (and therefore `make_matcher`).

I've updated to skip the generation of these functions rather than to fail completely.